File: c++/4.cpp
#include <iostream>
#include <vector>
#include <algorithm>

struct ListNode {
   int val;
   ListNode *next;
   ListNode(int x) : val(x), next(NULL) {}
};

bool isPalindrome(ListNode* head) {
   std::vector<int> vals;
   ListNode* current = head;
   while (current != NULL) {
       vals.push_back(current->val);
       current = current->next;
   }
   return vals == std::vector<int>(vals.rbegin(), vals.rend());
}

int main() {
   ListNode* head = new ListNode(1);
   head->next = new ListNode(2);
   head->next->next = new ListNode(2);
   head->next->next->next = new ListNode(1);

   std::cout << isPalindrome(head) << std::endl;
   return 0;
}

File: rust/4.rs
use std::collections::VecDeque;

#[derive(PartialEq, Eq, Clone, Debug)]
struct ListNode {
   val: i32,
   next: Option<Box<ListNode>>,
}

impl ListNode {
   fn new(val: i32) -> Self {
       ListNode {
           next: None,
           val,
       }
   }
}

fn is_palindrome(head: Option<Box<ListNode>>) -> bool {
   let mut vals = VecDeque::new();
   let mut current = head;
   while let Some(node) = current {
       vals.push_back(node.val);
       current = node.next;
   }
   vals == vals.iter().rev().cloned().collect::<VecDeque<_>>()
}

let mut head = Some(Box::new(ListNode::new(1)));
head.as_mut().unwrap().next = Some(Box::new(ListNode::new(2)));
head.as_mut().unwrap().next.as_mut().unwrap().next = Some(Box::new(ListNode::new(2)));
head.as_mut().unwrap().next.as_mut().unwrap().next.as_mut().unwrap().next = Some(Box::new(ListNode::new(1)));

println!("{}", is_palindrome(head));
